compiler/rustc_mir_transform/src/instcombine.rs

```diff
@@ -0,0 +1,405 @@
+        {
+            continue;
+        }
+
+        // We only do optimizations where there are no statements other than storage markers
+        // between the two statements. This is our proxy for alias analysis: if there are no
+        // statements between the creation of the temporary and its use, nothing can
+        // observe us eliding the temporary.
+        let statements_are_adjacent = &body.basic_blocks[temp_loc.block].statements
+            [temp_loc.statement_index + 1..final_loc.statement_index]
+            .iter()
+            .all(|s| {
+                matches!(
+                    &s.kind,
+                    StatementKind::Nop
+                        | StatementKind::StorageLive(_)
+                        | StatementKind::StorageDead(_)
+                )
+            });
+        if !statements_are_adjacent {
+            continue;
+        }
+
+        let cx = Context { tcx, local_decls: &body.local_decls, param_env };
+        for opt in &[
+            combine_place_projections,
+            combine_ref_deref,
+            combine_ptr_ptr_cast,
+            combine_ref_addressof,
+        ] {
+            let Some(new_statement) =
+                opt(&cx, temp_place, temp_rvalue, final_place, final_rvalue)
+            else { continue; };
+
+            invalidated_statements.insert(temp_loc);
+            invalidated_statements.insert(final_loc);
+
+            let statements = &mut body.basic_blocks_mut()[temp_loc.block].statements;
+
+            debug!(
+                "Combine:\n{:?}\n{:?}\n{:?}\n",
+                &statements[temp_loc.statement_index],
+                &statements[final_loc.statement_index],
+                &new_statement,
+            );
+
+            // We need to handle situations like this:
+            //
+            // _2 = _1 as *const u8;
+            // StorageDead(_1);
+            // StorageLive(_3);
+            // _3 = _2 as *const ();
+            //
+            // What we want to do is replace one of these assignments with
+            //
+            // _3 = _1 as *const ();
+            //
+            // But if we only replace one of the two assignments we analyze with our new
+            // statement, we will use a local outside of its liveness range.
+            // To deal with this, we remove the two original statements and all storage markers
+            // for locals in the new statement, and consider the original locations of these
+            // statements to be free slots in the block.
+            // Then we insert all our removed StorageLive statements into the first free slots,
+            // our StorageDead statements into the last slots, and our new statement somewhere in
+            // the middle.
+            // This ensures that we do not change the location of any statements that we have not
+            // optimized, which minimizes the amount of our analysis that we have invalidated.
```

Comment:
What happens if:
```
_2 = _1 as *const u8;
StorageDead(_4)
StorageLive(_4)
_3 = _2 as *const ();
```

Reply:
`_4` isn't mentioned in the combined statement, so those statements would be ignored. Did you mean `_3`?

Reply:
Indeed. My question is about the StorageDead-then-StorageLive getting swapped, so if that's the one triggering the behaviour, _3 it is.